SNV-callers: Add more background on the decisions made - docs and plots (#745)
AlexsLemonade/OpenPBTA-analysis: analyses/snv-callers/compare_snv_callers_plots.Rmd

plots these data in a series of plots to compare the consensus calls to the individual 
callers. 

*Notable Findings:*  

> - [VarDict calls more mutations](#upset-graph) than any of the other callers. 

Reviewer: ```suggestion
- [VarDict calls more mutations](#upset-graphs) than any of the other callers. 
```

> Many of these have [very low ](#vaf-distributions) suggesting that some may be sequencing errors. 

Reviewer: ```suggestion
Many of these have [very low Variant Allele Frequencies](#vaf-distributions) suggesting that they are more likely to be the result sequencing errors. 
```

> - Mutect2, Strelka2, and Lancet do have [plenty of mutations in common](#upset-graph). 

Reviewer: I feel like we can promote a bit of info toward the top here. Of course, it might not be accurate if the analysis changes, but I feel like it is worth making a point of
```suggestion
- Mutect2, Strelka2, and Lancet do have [a large number of mutations in common](#upset-graphs). 
- A 2 out of 3 consensus would increase the number of mutations called ~4X, but that set would be almost equivalent to a Mutect2 & Strelka2 consensus.
  -There are relatively few mutations found in only Mutect & Lancet or Strelka & Lancet (~125,000 vs 5,600,000 total mutations in at least 2 of 3 callers)
```

          plot_file_path = file.path(plots_dir, "pbta-upset-plot-transcipts-only.png"), 
          subset_vector = transcripts)
```

![Upset plot - transcripts only](plots/pbta-caller-comparison/pbta-upset-plot-transcipts-only.png)

Non transcript upset plot. 

```{r}
not_transcripts <- all_caller_df$Variant_Classification %in% c("5'Flank", "3'Flank", "IGR", "Intron") 

upset_png(detect_mat,
          plot_file_path = file.path(plots_dir, "pbta-upset-plot-not-transcipts-only.png"), 
          subset_vector = transcripts)
```

![Upset plot - transcripts only](plots/pbta-caller-comparison/pbta-upset-plot-not-transcipts-only.png)

## VAF distribution for each caller

Create consensus VAF based on strelka's VAFs and which mutations would be included

```{r}
# Make new VAF column based on strelka's VAFs
all_caller_df$VAF_consensus <- all_caller_df$VAF_strelka

# Figure out which mutations are not in the consensus
not_consensus <- dimnames(consensus_count)[[1]][which(consensus_count < 3)]

# Make non consensus mutations in consensus)VAF an NA
all_caller_df$VAF_consensus[not_consensus] <- NA
```

Make a long form VAF data.frame for exploring mutations VAFs from each caller. 

```{r}

> vaf_df <- all_caller_df %>% 

Reviewer: this is still going to include vardict, correct?

Author: Yes.

all_caller_df$VAF_consensus <- all_caller_df$VAF_strelka

# Figure out which mutations are not in the consensus
not_consensus <- dimnames(consensus_count)[[1]][which(consensus_count < 3)]

# Make non consensus mutations in consensus)VAF an NA
all_caller_df$VAF_consensus[not_consensus] <- NA
```

Make a long form VAF data.frame for exploring mutations VAFs from each caller. 

```{r}
vaf_df <- all_caller_df %>% 
  dplyr::select(index, dplyr::starts_with("VAF_"), Variant_Classification) %>%
  # Make long format
  tidyr::gather(key = "caller", value = "vaf", -index, -Variant_Classification) %>% 
  # Drop the `VAF_`
  dplyr::mutate(caller = gsub("VAF_", "", caller)) %>% 
  dplyr::filter(!is.na(vaf))
```

```{r}
# Make this plot
vaf_plot <- vaf_df %>%
  ggplot2::ggplot(ggplot2::aes(x = caller, y = vaf, color = caller)) +
  ggplot2::geom_violin() +
  ggplot2::theme_classic() +
  ggplot2::xlab("") +
  ggplot2::ylab("VAF") + 
  ggplot2::theme(legend.position = "none")

# Print out
vaf_plot
```

```{r}
# Save this plot
ggplot2::ggsave(plot = vaf_plot, 
                file.path(plots_dir, "pbta-vaf_distribution_plot.png"))
```

## VAF correlations

Correlate VAF's across callers.

> ```{r}

Reviewer: Can we hide warnings messages here? It gets awfully chatty!
```suggestion
```{r warning=FALSE}
```

cor_vaf <- all_caller_df %>% 
  dplyr::select(dplyr::starts_with("VAF_"), -VAF_consensus) %>% 

>   GGally::ggpairs(mapping = ggplot2::aes(alpha = 0.05)) 

Reviewer: `alpha` seems like it is being ignored here, or not getting its full due and I don't know enough to know why. Probably not worth worrying about.

Author: 🤷‍♀️ It does work if you look at the very fringes it just doesn't apply it to the lines only the fillings, and for PBTA there's just so many plots it doesn't matter anyway.

# Print out without progress bar
cor_vaf %>%
  print(progress = FALSE)
``` 

VAF's are pretty related across callers. 

> ```{r}

Reviewer: ```suggestion
```{r warning = FALSE}
```

# Remove the 0's
consensus_count <- consensus_count[which(consensus_count != 0)]
```

Print out the numbers for the three callers' consensus.

```{r}
cat(" At least 1 out of 3:", length(consensus_count), "\n",
    "At least 2 out of 3:", sum(consensus_count >= 2), "\n",
    "3 out of 3:", sum(consensus_count == 3))
```

## Upset graphs

Make the upset plot. 

```{r}
upset_png(detect_mat, 
          plot_file_path = file.path(plots_dir, "pbta-upset-plot.png"))
```

![Upset plot - all data](plots/pbta-caller-comparison/pbta-upset-plot.png)

Transcripts only upset plot. 

```{r}
transcripts <- !(all_caller_df$Variant_Classification %in% c("5'Flank", "3'Flank", "IGR", "Intron"))

upset_png(detect_mat, 
          plot_file_path = file.path(plots_dir, "pbta-upset-plot-transcipts-only.png"), 
          subset_vector = transcripts)
```

![Upset plot - transcripts only](plots/pbta-caller-comparison/pbta-upset-plot-transcipts-only.png)

Non transcript upset plot. 

```{r}
not_transcripts <- all_caller_df$Variant_Classification %in% c("5'Flank", "3'Flank", "IGR", "Intron") 

upset_png(detect_mat,
          plot_file_path = file.path(plots_dir, "pbta-upset-plot-not-transcipts-only.png"), 

>           subset_vector = transcripts)

Reviewer: ```suggestion
          subset_vector = not_transcripts)
```

>           subset_vector = transcripts)

Author: ```suggestion
          subset_vector = not_transcripts)
```

Curses.

```

> ![Upset plot - transcripts only](plots/pbta-caller-comparison/pbta-upset-plot-not-transcipts-only.png)

Reviewer: ```suggestion
![Upset plot - non-transcripts only](plots/pbta-caller-comparison/pbta-upset-plot-not-transcipts-only.png)
```

# Bring out the data.frame
all_caller_df <- all_caller %>%
  as.data.frame() %>%
  # We'll use this to keep track of the original rows/mutations
  tibble::rowid_to_column("index")

# Take a peek at what this looks like
head(all_caller_df)
```

## Get consensus numbers

Make a data.frame with only the VAFs and turn into detected or not. 

```{r}
detect_df <- all_caller_df %>% 
 # Bring over VAF columns and the index
 dplyr::select(dplyr::starts_with("VAF_"), index) %>% 
 # Turn into detected or not
 dplyr::mutate_at(dplyr::vars(dplyr::starts_with("VAF_")), function(x) !is.na(x))
```

Count up how many callers call each mutation but ignore VarDict. 

```{r}
consensus_count <- detect_df %>%
  # Ignore VarDict and index
  dplyr::select(-VAF_vardict, -index) %>%
  # Get the number of callers that call each mutation

>   dplyr::transmute(num = rowSums(.)) %>% 

Author: This step is quite a memory hog and will not run on even a 192 GB size AWS container for PBTA.

      TRUE ~ base_change
    )
  ) %>%

>   # Transmute VAFs to TRUE/FALSE

Reviewer: Is this comment what is happening here? I am confused by this set of steps, but I think mostly because `all_caller_df` probably had a lot of extra columns that could be trimmed away before joining. You used to do that, but now you don't?

Author: This comment is out of date. I've selected to just `Allele` and `Reference_Allele` since that's all I really needed from there.
